### ISSUE TITLE
Add new testcase where finality advances on alternate branch with lesser lib

### ISSUE THREAD

**Issue by greg7mdp** [AntelopeIO/spring#751]
Here is a whiteboard drawing of the [testcase](https://us06web.zoom.us/wb/doc/dcRtOGBzTtqnhsBoiR1ozw/p/81087274413) .

Same in a block comment:
```
/* -----------------------------------------------------------------------------------------------------
            Finality advancing past block claimed on alternate branch
            =========================================================

Time:        t1      t2      t3      t4      t5      t6      t7      t8
Blocks:
    B0 <---  B1 <--- B2 <--- B3 <-|- B4 <--- B5
                                  |
                                  \----------------- B6 <--- B7 <--- B8 <--- B9
QC claim:
           Strong          Strong  Strong  Strong  Strong   Weak   Strong  Strong
             B0              B1      B3      B4      B2      B6      B7      B8

Vote:                      Strong  Strong   Strong  Weak   Strong  Strong  Strong

                                                     ^
                                                     |
                                                 validating those weak votes on b2
                                                 would fail if nodes have received b4 and b5
                                                 which advanced lib to b3

    - Node D is isolated and has not seen B3, B4, and B5
    - it received B3 via push_block, (so it can make it its head and produce a child of B3), but has not
      received votes on b3 (only on b2), so b6 includes a strong QC on b2.
    - when b6 is pushed to A, B and C, qc validation should fail.
--------------------------------------------------------------------------------------------------------*/
```

Need a testcase similar as above, but where we avoid the problems that nodes which have seen block `B5` (to advance lib to `B3`) have voted on it (so locked on `B4`).

**Comment by greg7mdp** [comment]
@arhag This testcase does not work. After voting on `B5` (which makes `B3` final), the finalizers who voted on `B5` are locked on `B4`, and therefore cannot vote on `B6`:

- liveness check fails because: `B6' s core.latest_qc_block_timestamp() <  fsi.lock.timestamp` because `B2` timestamp < `B4` timestamp.
- safety check fails because `B6` does not extend `B4`.

**Comment by arhag** [comment]
@greg7mdp:

Try this out:

```
/* -----------------------------------------------------------------------------------------------------
            Finality advancing past block claimed on alternate branch
            =========================================================
Producer:    C       C       C       C       C       D       D       D       D
Timestamp:   t1      t2      t3      t4      t5      t6      t7      t8      t9
Blocks:
    B0 <---  B1 <--- B2 <--- B3 <-|- B4 <--- B5
                                  |
                                  \----------------- B6 <--- B7 <--- B8 <--- B9
QC claim:
           Strong  Strong  Strong  Strong  Strong  Strong  Strong   Weak   Strong
             B0      B0      B1      B3      B4      B1      B2      B7      B8

Votes:
  Node A:  Strong‡ Strong‡ Strong‡ Strong           Weak¹   Weak   Strong  Strong
  Node B:  Strong¹ Strong¹ Strong  Strong           Weak¹   Weak   Strong  Strong
  Node C:  Strong  Strong  Strong  Strong  Strong‡  Weak¹   Weak¹  Strong¹ Strong
  Node D:  Strong¹ Strong¹ Strong                  Strong  Strong  Strong  Strong

                                                             ^
                                                             |
                                             Validating the strong QC on B2 should
                                             not fail for nodes which receive B4 and
                                             B5 prior to B7 despite B5 advancing the
                                             fork DB root to B3.

Meaning of the superscripts and marks on the votes:
The vote on block b was delayed in reaching the node for the producer p scheduled 
for the block at the next time slot t after block b by enough that a block produced on time by
producer p for time slot t could not possibly utilize the vote in any QC the block could claim.
Furthermore, the delay is such that the earliest time slot at which producer p could
produce a block that utilizes the delayed vote is the time slot (t + d) where ...
¹ ... d = 1.
‡ ... d is infinite meaning the vote may never be received by producer p.

--------------------------------------------------------------------------------------------------------*/
```

The connectivity between the nodes is as follows:

```
A---D  
|\  |
| \ |
|  \|
C---B 
```

Requirements to make this work:
+ D must receive less than three votes for B2 prior to producing B6.
+ D must receive three strong votes for B2 prior to producing B7.
+ D must produce B8 before it receives B4.
+ A and B must receive B4 before B6, receive B6 before B5, and receive B5 before B7.

Timeline:

1. Node C produces and broadcasts B1 which has a strong QC on B0.
2. Nodes A, B, C, and D receive B1, vote on it, and broadcast their strong votes. Each node immediately receives its own strong vote.
3. Node A is isolated from the other nodes.
4. Node C produces and broadcasts B2 which repeats the strong QC claim on B0.
5. Node C receives B2, votes on it, and broadcasts its strong vote. Node C immediately receives its own strong vote.
6. Node C receives the strong votes on B1 from nodes B and D and forms a valid strong QC for B1 in its fork database.
7. Node D receives the strong votes on B1 from nodes B and C and forms a valid strong QC for B1 in its fork database.
8. Node C produces and broadcasts B3 which has a strong QC on B1.
9. Node C receives B3, votes on it, and broadcasts its strong vote. Node C immediately receives its own strong vote.
10. Nodes B and D receive B2, vote on it, and broadcast their strong votes. Each of the two nodes immediately receives its own strong vote.
11. Nodes B and D receive B3, vote on it, and broadcast their strong votes. Each of the two nodes immediately receives its own strong vote.
12. Node C receives the strong votes on B2 from nodes B and D and forms a valid strong QC for B2 in its fork database.
13. Node C receives the strong votes on B3 from nodes B and D and forms a valid strong QC for B3 in its fork database.
14. Node D receives the strong vote on B2 from node B.
15. Node D is isolated from the other nodes.
16. Node C produces and broadcasts B4 which has a strong QC on B3.
17. Nodes B and C receive B4 which advances LIB on their nodes to B1. They each vote on it, and broadcasts their strong votes. Each of the two nodes immediately receives its own strong vote.
18. Node D produces and broadcasts B6 one second early (due to clock differences) which has B3 as its parent block and repeats the strong QC claim on B1.
19. Node D receives B6, votes on it, and broadcasts its strong vote. Node D immediately receives its own strong vote.
20. Nodes A and D establish connection to each other but remain isolated from nodes B and C.
21. Node A receives B2, votes on it, and broadcasts its strong vote. Node A immediately receives its own strong vote.
22. Node D receives the strong vote on B2 from node A and forms a valid strong QC for B2 in its fork database.
23. Node A receives B3, votes on it, and broadcasts its strong vote. Node A immediately receives its own strong vote.
24. Node D is again isolated from the other nodes including node A.
25. Node A establishes connection with nodes B and C. But D remains isolated from the other three nodes.
26. Node A receives B4 which advances LIB on its node to B1. A votes on it, and broadcasts its strong vote. Node A immediately receives its own strong vote.
27. Node C receives the strong votes on B4 from nodes A and B and forms a valid strong QC for B4 in its fork database.
28. Node C is isolated from the other nodes.
29. Node C produces and broadcasts B5 which has a strong QC on B4.
30. Node C receives B5, votes on it, and broadcasts its strong vote. Node C immediately receives its own strong vote.
31. Node D produces and broadcasts B7 one second early (due to clock differences) which has a strong QC on B2.
32. Node D receives B7, votes on it, and broadcasts its strong vote. Node D immediately receives its own strong vote.
33. Node D establishes connection with nodes A and B. But C remains isolated from the other three nodes.
34. Nodes A and B receive B6 which is added to the fork database but does not become the new best head since B4 has a better QC claim. A and B vote on B6, and broadcast their weak votes.
35. Node D receives the weak votes on B6 from nodes A and B and forms a valid weak QC for B6 in its fork database.
36. Node C establishes connection with nodes A and B and (indirectly) to node D.
37. Nodes A and B receive B5 which advances LIB on their nodes to B3. A and B do not vote on B5 because their latest vote was on block B6 which has a more recent timestamp than B5.
38. Node C receives B6 which is added to the fork database but does not become the new best head since B5 has a better QC claim. C votes on B6, and broadcasts its weak vote.
39. Nodes A and B receive B7 which is added to the fork database but does not become the new best head since B5 has a better QC claim. A and B vote on B7, and broadcast their weak votes.
40. Node D receives the weak votes on B7 from nodes A and B and forms a valid weak QC for B7 in its fork database.
41. Node C receives B7 which is added to the fork database but does not become the new best head since B5 has a better QC claim. C votes on B7, and broadcasts its weak vote.
42. Node D produces and broadcasts B8 one second early (due to clock differences) which has a weak QC on B7.
43. Node D receives B8, votes on it, and broadcasts its strong vote. Node D immediately receives its own strong vote.
44. Node D receives B4 which advances LIB on its node to B1. D does not vote on B4 because its latest vote was on block B8 which has a more recent timestamp than B4.
45. Nodes A and B receive B8 which becomes their new best head. They vote on it, and broadcast their strong votes.
46. Node D receives B5 which advances LIB on its node to B3. D does not vote on B5 because its latest vote was on block B8 which has a more recent timestamp than B5.
47. Node D receives the strong votes on B8 from nodes A and B and forms a valid strong QC on B8 in its fork database.
48. Node C receives B8 which becomes its new best head. C votes on it, and broadcasts its strong vote.
49. Node D produces and broadcasts B9 one second early (due to clock differences) which has a strong QC on B8.
50. Nodes A, B, C, and D receive B9 which advances LIB on their nodes to B7. They vote on it, and broadcast their strong votes. Each node immediately receives its own strong vote.

Step 39 in the timeline above is the step that would demonstrate the bug. Nodes A and B would have the blocks up to and including B5 and B6 just prior to step 39. This means their LIB would be B3. When they receive block B7, it would successfully link with block B6 in their fork database. However, block B7 has a QC for B2 which is older than LIB. So block B2 could not be found in the fork database to validate the QC included in block B7. Instead, the `block_header_state` of B6 would need to have all the information needed to validate the QC for B2 which is include in block B7.

**Comment by greg7mdp** [comment]
@arhag before node D is isolated from the other nodes (step 15),  it received b3 and 3 votes on it (from B, C and D), so when it produces b6 at step 18, b6 includes a strong QC on b3.

**Comment by arhag** [comment]
> before node D is isolated from the other nodes (step 15), it received b3 and 3 votes on it (from B, C and D), so when it produces b6 at step 18, b6 includes a strong QC on b3.

@greg7mdp: Oh well, so I guess I did rely on delays of votes after all. All right, so we will need the new capability you added to the Savanna cluster to delay votes.

**Comment by arhag** [comment]
@greg7mdp:

Actually, it may still not be necessary to introduce vote propagation delays.

We just need to make sure that node D receives block B3 but does not receive 3 votes on B3. It is guaranteed to receive its own vote on B3. And unless we introduce vote propagation delays, it will also receive C's vote on B3 since C produce B3. But we could guarantee that node D does not receive B's vote on B3 by making sure that node D receives block B3 and then gets isolated before node B receives block B3.

**Comment by greg7mdp** [comment]
> Notice that the only significant differences between the primed timeline and the unprimed timeline are nodes A, B, and C do not vote on B7, B8 has a QC for B6 (rather than B7), and B9 causes LIB on all nodes to advance to B6 (rather than B7).
> 
> You should be able to verify this primed timeline behavior occurs with #778 remaining unfixed. And then you should be able to show that with #778 fixed, the unprimed timeline behavior occurs instead.

I have finished reproducing the two scenarios. As expected because #694 is fixed,  A and B can validate the received block B7 at step 41.

But even when I merge the fix for #778, A and B still not vote on b7. This is because when pushing b6 and b7 to A, the b6/b7 branch does *not* become the best head, and as a result we do not do a fork switch and b6/b7 are not applied  (and marked `valid` in fork_db).

When considering to vote on b7, we see that it claims b6, but b6 not being marked valid in fork_db we decline to vote on it.
